espnet2/speechlm/core_lm/ar.py

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python3
+
+# Copyright 2024 Jinchuan Tian
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+# Implementation of naive langauge model with codec interleave
+
+from typing import Tuple, Dict
+import torch
+import logging
+
+from espnet2.speechlm.core_lm.abs_core_lm import AbsCoreLM, SpeechLMInferenceOptions
+from espnet2.speechlm.module.module import (
+    TransformerLayer,
+    PositionalEncoding,
+)
+from espnet2.speechlm.net_utils import (
+    ce_loss,
+    install_kv_cache_hook,
+    logits_to_tokens,
+)
+
+
+class ARLM(AbsCoreLM):
+    def __init__(
+        self,
+        vocab_size: int,
+        nq: int,
+        share_emb: bool = False,
+        pos_enc: str = None,
+        att_unit: int = 256,
+        head: int = 2,
+        unit: int = 1024,
+        layer: int = 4,
+        dropout_rate: float = 0.0,
+        positional_dropout_rate: float = 0.0,
+        attention_dropout_rate: float = 0.0,
+    ):
+        super(ARLM, self).__init__()
+
+        if pos_enc == "sinusoidal":
+            pos_enc_class = PositionalEncoding
+        else:
+            raise ValueError(f"unknown pos-enc option: {pos_enc}")
+
+        self.emb = torch.nn.Embedding(vocab_size, att_unit)
+        self.lm_head = torch.nn.Linear(att_unit, vocab_size * nq, bias=False)
+        if share_emb:
+            raise ValueError("Embedding cannot be shared")
+
+        self.decoders = torch.nn.ModuleList(
+            [
+                TransformerLayer(
+                    att_unit=att_unit,
+                    head=head,
+                    unit=unit,
+                    dropout_rate=dropout_rate,
+                    attention_dropout_rate=attention_dropout_rate,
+                    causal=True,
+                    cross_attention=False,
+                )
+                for _ in range(layer)
+            ]
+        )
+        self.post_ln = torch.nn.LayerNorm(att_unit)
+        self.pos_enc = pos_enc_class(att_unit, positional_dropout_rate)
+
+        self.nq = nq
+
+    def forward(
+        self,
+        dec_seq: torch.Tensor,
+        dec_seq_lengths: torch.Tensor = None,
+        enc_seq: torch.Tensor = None,
```

Comment:
what is the enc_seq here for?

Reply:
enc_seq is reserved for encoder-decoder structure. Maybe we can remove it at this moment.

Reply:
I see. I think we can safely remove it (or add TODO to that for corresponding occurrence